src/pecuzal_method.jl

```diff
@@ -411,3 +418,177 @@ function get_maxima(s::Vector{T}) where {T}
     end
     return maximas, maximas_idx
 end
+
+
+"""
+    uzal_cost_pecuzal(Y1::Dataset, Y2::Dataset, Tw; kwargs...) → L_decrease
+This function is based on the functionality of [`uzal_cost`](@ref), here
+specifically tailored for the needs in the PECUZAL algorithm.
+Compute the L-statistics `L1` and `L2` for the input datasets `Y1` and `Y2` for
+increasing time horizons `T = 1:Tw`. For each `T`, compute `L1` and `L2` and
+decrease `L_decrease = L2 - L1`. If `L_decrease` is a negative value, then `Y2`
+can be regarded as a "better" reconstruction that `Y1`. Break, when `L_decrease`
+reaches the 1st local minima, since this will typically also be the global
+minimum. Return the according minimum `L_decrease`-value.
+
+## Keyword arguments
+
+* `K = 3`: the amount of nearest neighbors considered, in order to compute σ_k^2
+  (read algorithm description).
+  If given a vector, minimum result over all `k ∈ K` is returned.
+* `metric = Euclidean()`: metric used for finding nearest neigbhors in the input
+  state space trajectory `Y.
+* `w = 1`: Theiler window (neighbors in time with index `w` close to the point,
+  that are excluded from being true neighbors). `w=0` means to exclude only the
+  point itself, and no temporal neighbors.
+"""
+function uzal_cost_pecuzal(Y::Dataset{D, ET}, Y_trial::Dataset{DT, ET}, Tw::Int;
+        K::Int = 3, w::Int = 1, metric = Euclidean()
+    ) where {D, DT, ET}
+
+    @assert DT == D+1
+    @assert Tw ≥ 0
+
+    NNN = length(Y_trial)-1;
+    # preallocation for 1st dataset
+    ϵ² = zeros(NNN)             # neighborhood size
+    E²_avrg = zeros(NNN)        # averaged conditional variance
+    E² = zeros(NNN, Tw)         # conditional variance
+    ϵ_ball = zeros(ET, K+1, D)  # epsilon neighbourhood
+    u_k = zeros(ET, D)          # center of mass
+    # preallocation for 2nd dataset
+    ϵ²_trial = zeros(NNN)             # neighborhood size
+    E²_avrg_trial = zeros(NNN)        # averaged conditional variance
+    E²_trial = zeros(NNN, Tw)         # conditional variance
+    ϵ_ball_trial = zeros(ET, K+1, DT) # epsilon neighbourhood
+    u_k_trial = zeros(ET, DT)         # center of mass
+
+    dist_former = 9999999 # intial L-decrease
+
+    # loop over each time horizon
+    cnt = 1
+    for T = 2:Tw    # start at 2 will eliminate results for noise
+        NN = length(Y_trial)-T;
+        ns = 1:NN
+
+        vs = Y[ns] # the fiducial points in the data set
+        vs_trial = Y_trial[ns] # the fiducial points in the data set
+
+        vtree = KDTree(Y[1:NN], metric)
```

Comment:
Instead of re-creating a tree everytime, perhaps it is better to have a tree with all points and simply exclude all points with indices `i > NN-T` from the neighbor search. Since arbitrary skip functions can be given to the algorithm this will increase performance significantly, since now the tree will be made 1 time instead of e.g. 20 (default Tw)